### PR TITLE
[GH-9] Refactored proof generation and verification, implemented proof lookups

### PR DESCRIPTION
**Pull request by grzegorz225**
This PR resolves GH-9 and GH-11

### CODE REVIEW

**Comment by DanielaIvanova** [comment]
If we insert an account in the tree the root hashes are the same as epoch.

File: lib/proof.ex [hunk 1]
```diff
@@ -2,179 +2,256 @@ defmodule MerklePatriciaTree.Proof do
...
-        case Enum.at(branches, nibble) do
-          [] ->
-            {nil, proof}
+  @spec int_construct_proof(decoded_node(), Trie.t(), list(), Trie.t()) :: {Trie.val() | nil, Trie.t()} | decode_node_error()
+  defp int_construct_proof(:empty, _, _, proof), do: {nil, proof}
 
-          node_hash when is_binary(node_hash) and byte_size(node_hash) == 32 ->
-            node = insert_proof_db(node_hash, trie.db, proof)
-            construct_proof({node, trie}, rest, proof)
+  defp int_construct_proof({:branch, branches}, _, [], proof) do
```
Comment (thepiwo) [inline lib/proof.ex]: what is `int_` for?

Comment (thepiwo) [reply lib/proof.ex]: maybe rename to something more semantic correct or full name

File: lib/proof.ex [hunk 2]
```diff
@@ -2,179 +2,256 @@ defmodule MerklePatriciaTree.Proof do
...
+    {List.last(branches), proof}
+  end
 
-          node_hash ->
-            construct_proof({node_hash, trie}, rest, proof)
-        end
+  defp int_construct_proof({:leaf, [], v}, _, [], proof) do
```
Comment (thepiwo) [inline lib/proof.ex]: rename v to value I assume

File: lib/proof.ex [hunk 3]
```diff
@@ -2,179 +2,256 @@ defmodule MerklePatriciaTree.Proof do
...
-    ## TODO : handle when there is no value into the db
-    {:ok, node} = read_from_db(proof.db, hash)
-
-    case decode_node(ExRLP.decode(node), nil, proof) do
-      :error -> false
-      node -> int_verify_proof(Helper.get_nibbles(key), node, value, proof)
+  @spec lookup_proof(Trie.key(), binary(), Trie.t()) :: Trie.value() | proof_verification_error()
+  def lookup_proof(key, root_hash, proof) do
+    case verify_proof(key, :lookup, root_hash, proof) do
+      {:error, {:bad_val, val}} ->
```
Comment (thepiwo) [inline lib/proof.ex]: rename to value to be consistent

File: lib/proof.ex [hunk 4]
```diff
@@ -2,179 +2,256 @@ defmodule MerklePatriciaTree.Proof do
...
-    {:ext, prefix, ExRLP.decode(rlp_node)}
+  @spec decode_node(list()) :: {:ok, {atom(), list()}} | {:error, :invalid_node}
+  defp decode_node(branch) when is_list(branch) and length(branch) == 17 do
+    {:ok, {:branch, branch}}
   end
 
-  defp build_ext(prefix, hash, _trie, _proof) do
-    IO.inspect("3")
-    {:ext, prefix, hash}
+  defp decode_node([hp_k, v]) do
```
Comment (thepiwo) [inline lib/proof.ex]: rename hp_k and v to semantic correct names

**Comment by grzegorz225** [comment]
@thepiwo I fixed the inconsistencies in the code.

**Pull request merged**